python/ray/spark/__init__.py

```diff
@@ -0,0 +1,313 @@
+    ray_head_port = get_safe_port(ray_head_hostname)
+
+    logging.info(f"Ray head hostanme {ray_head_hostname}, port {ray_head_port}")
+
+    num_spark_task_cpus = int(spark.sparkContext.getConf().get("spark.task.cpus", "1"))
+    num_spark_task_gpus = int(spark.sparkContext.getConf().get("spark.task.resource.gpu.amount", "0"))
+
+    ray_worker_heap_mem_bytes, ray_worker_object_store_mem_bytes = get_avail_mem_per_ray_worker(spark)
+
+    ray_exec_path = os.path.join(os.path.dirname(sys.executable), "ray")
+
+    ray_node_log_dir = os.path.join(ray_node_log_dir, f"cluster-{ray_head_hostname}-{ray_head_port}")
+
+    logging.warning(f"You can check ray head / worker nodes logs under local disk path {ray_node_log_dir}")
+    if is_in_databricks_runtime() and not ray_node_log_dir.startswith("/dbfs"):
+        logging.warning(
+            "We recommend you to set `ray_node_log_root_path` argument to be a path under '/dbfs/', "
+            "because for all spark cluster nodes '/dbfs/' path is mounted with a shared disk, "
+            "so that you can check ray worker logs on spark driver node."
+        )
+
+    os.makedirs(ray_node_log_dir, exist_ok=True)
+
+    ray_temp_dir = os.path.join(ray_temp_dir, f"cluster-{ray_head_hostname}-{ray_head_port}")
+    ray_head_temp_dir = os.path.join(ray_temp_dir, "head")
+    os.makedirs(ray_head_temp_dir, exist_ok=True)
+
+    ray_head_node_cmd = [
+        ray_exec_path,
+        "start",
+        f"--temp-dir={ray_head_temp_dir}",
+        "--block",
+        "--head",
+        "--disable-usage-stats",
+        f"--port={ray_head_port}",
+        "--include-dashboard=false",
+        f"--num-cpus=0",  # disallow ray tasks scheduled to ray head node.
+        # limit the memory usage of head node because no task running on it.
+        f"--memory={128 * 1024 * 1024}",
+        # limit the object store memory usage of head node because no task running on it.
+        f"--object-store-memory={128 * 1024 * 1024}",
+        *_convert_ray_node_options(head_options)
+    ]
+
+    logging.info(f"Start Ray head, command: {' '.join(ray_head_node_cmd)}")
+
+    with open(os.path.join(ray_node_log_dir, "ray-head.log"), "w", buffering=1) as head_log_fp:
+        ray_head_proc = exec_cmd(
+            ray_head_node_cmd,
+            synchronous=False,
+            capture_output=False,
+            stream_output=False,
+            stdout=head_log_fp,
+            stderr=subprocess.STDOUT,
+        )
+
+    # wait ray head node spin up.
+    wait_ray_node_available(
+        ray_head_hostname, ray_head_port, 40,
+        error_on_failure="Start Ray head node failed!"
+    )
+
+    logging.info("Ray head node started.")
+
+    def ray_cluster_job_mapper(_):
+        from pyspark.taskcontext import BarrierTaskContext
+
+        context = BarrierTaskContext.get()
+        context.barrier()
+        task_id = context.partitionId()
+
+        # TODO: remove temp dir when ray worker exits.
```

Comment:
@ericl @jjyao 

Could you help add an option for `ray start` script to make ray node delete temp directory when ray node is killed ?

Reply:
I don't think this is something we could offer except as best effort, which suggests it isn't the right mechanism for cleanup. Instead, why not set the temp dir to a known location and remove that in a wrapper script after the Ray worker exits?

Reply:
@ericl Yes. Make sense. I use a wrapper script and it register a SIGTERM handler and in handler it deletes the temp dir.